### PR TITLE
feat: composable workflows and orchestrator execution mode support

### LINKED ISSUE #20: Hierarchical Workflow Management with Rollback Support
### Feature to Enhance

Enhance the workflow engine to support hierarchical workflows, allowing addition and deletion of sub-workflows (each containing steps), sequential execution, and robust rollback on error.

## Sub-Workflow Management
- Ability to add sub-workflows within a parent workflow.
- Each sub-workflow can contain multiple steps.

## Sequential Execution
- Executing a workflow should run all its sub-workflows in defined order.
- Each sub-workflow executes its steps sequentially.

## Rollback on Error
If any step or sub-workflow fails, initiate rollback.
- Rollback all sub-workflows in reverse order.
- Each sub-workflow rolls back its steps in reverse order.

## Acceptance Criteria:
- API or interface to build workflows with sub-workflows.
- Execution engine runs sub-workflows and their steps in order.
- On error, rollback is triggered as described.
- Rollback operations are idempotent and handle partial failures gracefully.

### Business Need

This feature will provide flexible, reliable workflow orchestration with strong error recovery, supporting complex automation scenarios.

### Additional Context

_No response_

### LINKED ISSUE #22: Replace Choreography Execution Model with Orchestrator for Centralized Workflow Control
Currently, the workflow engine uses a choreography execution model, where each step is responsible for invoking the next step or triggering rollback. This decentralized approach can make it difficult to enforce business rules, handle errors consistently, and monitor workflow progress.

### Proposal:
Replace the choreography execution model with an orchestrator execution model. In this model, a central orchestrator manages the execution order, error handling, and rollback of all steps and sub-workflows.

### Benefits:

- Centralized Control: Easier to enforce business logic and manage execution flow.
- Consistent Error Handling: Centralized error, retry, and compensation logic.
- Simplified Composition: Easier to compose and nest workflows.
- Improved Observability: Single point for logging, reporting, and monitoring.

### Acceptance Criteria:

- The orchestrator (aka `Workflow` execution) manages all step and sub-workflows execution.
- Error handling and rollback are performed centrally.
- Existing tests pass; new tests cover orchestrator scenarios.
- Documentation and examples are updated to reflect the new model.

### Additional Context:
This change will enable more robust, maintainable, and observable workflow execution, supporting future extensibility and advanced features.

### Plan of Action
- Add `Step` interface to `Workflows` in order to support #20 
- Implement `Forward` and `Reverse` methods of `Step` interface for `Workflows` to implement orchestrator execution model

### PR DESCRIPTION
**Pull request by leninmehedy**
## Description

This PR introduces composable workflows to the automa framework, enabling the construction of complex automation processes from reusable, modular workflow components. The design allows individual workflows to be defined, registered, and then composed together, supporting both step-level and workflow-level composition.

## Design Considerations
- Modularity: Each workflow and step is defined independently, promoting separation of concerns and easier maintenance.

- Reusability: Common setup or deployment logic can be encapsulated in a workflow and reused across multiple higher-level workflows.

- Granular Control: Each workflow can specify its own rollback mode and error handling strategy, allowing for fine-tuned control over execution and recovery.

- Registry Pattern: Workflows and steps are registered in a central registry, making them discoverable and composable by name.

- Extensibility: New workflows or steps can be added without modifying existing logic, supporting scalable automation as requirements grow.

### Benefits
- Reduced Duplication: Shared logic is implemented once and reused, minimizing code duplication and potential for errors.
- Improved Maintainability: Changes to a common workflow propagate automatically to all composed workflows, simplifying updates and bug fixes.
- Enhanced Testability: Individual workflows and steps can be tested in isolation, improving reliability and confidence in automation.
- Flexible Composition: Workflows can be nested or combined in various ways to address different automation scenarios, from simple setups to complex CI/CD pipelines.
- Consistent Rollback: Each workflow manages its own rollback behavior, ensuring predictable and safe recovery from failures.

### Example
A local development environment setup workflow can be composed with other workflows (e.g., CI/CD setup), enabling teams to build robust, repeatable automation pipelines with minimal effort. See `docs/example`
<hr></hr>
This design lays the foundation for scalable, maintainable, and robust automation using the automa framework.

* Closes #20 
* Closes #22 
* Related to #

### Pull request (PR) checklist

* \[ ] This PR added tests (unit, integration, and/or end-to-end)
* \[ ] This PR updated documentation
* \[ ] This PR added no TODOs or commented out code
* \[ ] This PR has no breaking changes
* \[ ] Any technical debt has been documented as a separate issue and linked to this PR
* \[ ] Any `go.mod` and/or `go.sum` changes have been explained to and approved by a repository manager
* \[ ] All related issues have been linked to this PR
* \[ ] All changes in this PR are included in the description
* \[ ] When this PR merges the commits will be squashed and the title will be used as the commit message, the 'commit
  message guidelines' below have been followed

### Testing

* \[ ] This PR added unit tests
* \[ ] This PR added integration/end-to-end tests
* \[ ] These changes required manual testing that is documented below
* \[ ] Anything not tested is documented

The following manual testing was done:

* TBD

The following was not tested:

* TBD

<details>
<summary>
Commit message guidelines
</summary>
We use 'Conventional Commits' to ensure that our commit messages are easy to read, follow a consistent format, and for automated release note generation. Please follow the guidelines below when writing your commit messages:

1. BREAKING CHANGE: a commit that has a footer BREAKING CHANGE:, or appends a ! after the type/scope, introduces a
   breaking API change (correlating with MAJOR in Semantic Versioning). A BREAKING CHANGE can be part of commits of any
   type. NOTE: currently breaking changes will only bump the MAJOR version.
2. The title is prefixed with one of the following:

| Prefix    | Description                                         | Semantic Version Update | Captured in Release Notes |
|-----------|-----------------------------------------------------|-------------------------|---------------------------|
| feat:     | a new feature                                       | MINOR                   | Yes                       |
| fix:      | a bug fix                                           | PATCH                   | Yes                       |
| perf:     | performance                                         | PATCH                   | Yes                       |
| refactor: | code change that isn't feature or fix               | none                    | No                        |
| test:     | adding missing tests                                | none                    | No                        |
| docs:     | changes to documentation                            | none                    | Yes                       |
| build:    | changes to build process                            | none                    | No                        |
| ci:       | changes to CI configuration                         | none                    | No                        |
| style:    | formatting, missing semi-colons, etc                | none                    | No                        |
| chore:    | updating grunt tasks etc; no production code change | none                    | No                        |

</details>

### CODE REVIEW

File: automa_steps/step_default.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	if s.onPrepare != nil {
+		c, err := s.onPrepare(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		s.ctx = c
+	}
+
+	return s.ctx, nil
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply automa_steps/step_default.go:43]: We need to initialize with ctx instead of setting it nil. Fixed and pushed.

File: workflow.go [hunk 2]
```diff
@@ -1,141 +1,158 @@
...
+			// create failure report for the failed step
+			failureReport := StepFailureReport(step.Id(), WithActionType(ActionExecute), WithStartTime(startTime), WithError(err))
+			stepReports = append(stepReports, failureReport)
+
+			// Perform rollback for all executed steps up to the current one
+			rollbackReports := w.rollbackFrom(ctx, index)
+			if len(rollbackReports) != len(stepReports) {
+				return nil, errorx.IllegalState.New("mismatched rollback reports and step reports lengths, "+
+					"rollback reports: %d, step reports: %d", len(rollbackReports), len(stepReports))
+			}
```
[the remark this answers is not in the document]
Comment (leninmehedy) [reply workflow.go]: Add check for rollback mode
